### PR TITLE
HDDS-1065. OM and DN should persist SCM certificate as the trust root. Contributed by Ajay Kumar.

### PR DESCRIPTION
**Pull request by ajayydv**
(cherry picked from commit 7254bf06e66deaf4dd9d00e65fc8894bd869a797)

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 32 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 62 | Maven dependency ordering for branch |
| +1 | mvninstall | 395 | trunk passed |
| +1 | compile | 202 | trunk passed |
| +1 | checkstyle | 53 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 823 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 122 | trunk passed |
| 0 | spotbugs | 235 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 422 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 25 | Maven dependency ordering for patch |
| +1 | mvninstall | 377 | the patch passed |
| +1 | compile | 195 | the patch passed |
| +1 | javac | 195 | the patch passed |
| -0 | checkstyle | 23 | hadoop-ozone: The patch generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0) |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 604 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 116 | the patch passed |
| +1 | findbugs | 422 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 137 | hadoop-hdds in the patch failed. |
| -1 | unit | 1239 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 41 | The patch does not generate ASF License warnings. |
| | | 5446 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.client.rpc.TestWatchForCommit |
|   | hadoop.ozone.client.rpc.TestBlockOutputStreamWithFailures |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-834/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/834 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 03cf50d376b3 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 24c53e0 |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-834/1/artifact/out/diff-checkstyle-hadoop-ozone.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-834/1/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-834/1/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-834/1/testReport/ |
| Max. process+thread count | 4278 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/common hadoop-hdds/container-service hadoop-ozone/integration-test hadoop-ozone/ozone-manager U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-834/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/CertificateClient.java [hunk 1]
```diff
@@ -135,10 +135,11 @@ boolean verifySignature(byte[] data, byte[] signature,
    *
    * @param pemEncodedCert        - pem encoded X509 Certificate
    * @param force                 - override any existing file
+   * @param caCert                - Is CA certificate.
    * @throws CertificateException - on Error.
    *
    */
-  void storeCertificate(String pemEncodedCert, boolean force)
+  void storeCertificate(String pemEncodedCert, boolean force, boolean caCert)
```
Comment (xiaoyuyao) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/CertificateClient.java]: Adding a new interface method storeCACertificate() can avoid many unrelated changes.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 34 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 1 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 65 | Maven dependency ordering for branch |
| +1 | mvninstall | 533 | trunk passed |
| +1 | compile | 267 | trunk passed |
| +1 | checkstyle | 65 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 817 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 141 | trunk passed |
| 0 | spotbugs | 308 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 507 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 27 | Maven dependency ordering for patch |
| +1 | mvninstall | 484 | the patch passed |
| +1 | compile | 268 | the patch passed |
| +1 | cc | 268 | the patch passed |
| +1 | javac | 268 | the patch passed |
| +1 | checkstyle | 68 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 608 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 135 | the patch passed |
| +1 | findbugs | 479 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 144 | hadoop-hdds in the patch failed. |
| -1 | unit | 1236 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 42 | The patch does not generate ASF License warnings. |
| | | 6051 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.client.rpc.TestBlockOutputStreamWithFailures |
|   | hadoop.ozone.client.rpc.TestCommitWatcher |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-834/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/834 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle cc |
| uname | Linux 9c5dc63198a5 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 1cb2eb0 |
| Default Java | 1.8.0_212 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-834/3/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-834/3/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-834/3/testReport/ |
| Max. process+thread count | 4975 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/common hadoop-hdds/container-service hadoop-ozone/integration-test hadoop-ozone/ozone-manager U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-834/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 34 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 3 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 61 | Maven dependency ordering for branch |
| +1 | mvninstall | 571 | trunk passed |
| +1 | compile | 276 | trunk passed |
| +1 | checkstyle | 73 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 846 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 136 | trunk passed |
| 0 | spotbugs | 306 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 507 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 31 | Maven dependency ordering for patch |
| +1 | mvninstall | 506 | the patch passed |
| +1 | compile | 278 | the patch passed |
| +1 | cc | 278 | the patch passed |
| +1 | javac | 278 | the patch passed |
| +1 | checkstyle | 77 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 655 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 145 | the patch passed |
| +1 | findbugs | 531 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 149 | hadoop-hdds in the patch failed. |
| -1 | unit | 926 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 58 | The patch does not generate ASF License warnings. |
| | | 6018 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-834/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/834 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle cc |
| uname | Linux b21d89766c96 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 1cb2eb0 |
| Default Java | 1.8.0_212 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-834/4/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-834/4/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-834/4/testReport/ |
| Max. process+thread count | 5404 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/common hadoop-hdds/container-service hadoop-ozone/integration-test hadoop-ozone/ozone-manager U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-834/4/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by xiaoyuyao** [comment]
+1. Thanks @ajayydv for the contribution.
